numpy/core/src/umath/loops.c.src

```diff
@@ -1656,6 +1680,78 @@ FLOAT_@func@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSE
  * #CHK = HAVE_ATTRIBUTE_TARGET_AVX512F_WITH_INTRINSICS, HAVE_ATTRIBUTE_TARGET_AVX2_WITH_INTRINSICS#
  */
 
+/**begin repeat1
+ *  #TYPE = FLOAT, DOUBLE#
+ *  #type = npy_float, npy_double#
+ *  #typesub = f, #
+ */
+
+NPY_NO_EXPORT NPY_GCC_OPT_3 void
+@TYPE@_sqrt_@isa@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(data))
+{
+    if (!run_unary_@isa@_sqrt_@TYPE@(args, dimensions, steps)) {
+        UNARY_LOOP {
+            const @type@ in1 = *(@type@ *)ip1;
+            *(@type@ *)op1 = npy_sqrt@typesub@(in1);
+        }
+    }
```

Comment:
Is the compiler able to generate the avx code automatically if you use
```
if (IS_OUTPUT_BLOCKABLE_UNARY(sizeof(@type@), @REGISTER_SIZE@)) {
    UNARY_LOOP { ... }
}
else {
    // as above
    UNARY_LOOP { ... }
}
```
We use this trick in all sorts of places today to encourage it to generate optimized code.

Reply:
I tried several options with GCC-9.2 and found the following: 

- Any compiler generated vectorized loop for floating point seems to require extra compiler options like -ffast-math (see https://gcc.gnu.org/projects/tree-ssa/vectorization.html#using) . Here is the code for an example of the sqrt loop [with ](https://godbolt.org/z/Ctlf4S)and [without ](https://godbolt.org/z/ElB339) this option. There are several problems with this path: (1) -ffast-math obviously should not be used as a global compile option and (2) the code generated with this option ends up using a combination of `vrsqrt14ps `and `vmulps `instruction to compute square root which is neither accurate nor fast (`vrsqrt14ps  `is only accurate up to the 6th decimal place and I have no idea why even the latest GCC wont use a simple `vsqrtps ` instruction instead!)

- The other problem is, no matter what option I try, I could not get GCC to vectorize the strided array case (see an example [here](https://godbolt.org/z/McQnDf)). Even if somehow we were able to properly vectorize the case where stride = 1, as far as I know, we cannot auto-vectorize for general strided arrays.

Reply:
I finally learnt why gcc wont use `vsqrtps`! `vrsqrt14ps `instruction is 1-3 cycles, where as `vsqrtps `is > 14 cycles. So its basically faster to compute `invsqrt `and then correct it with one step of newton raphson than to compute an accurate sqrt directly. -ffast-math obviously chooses speed over accuracy. This logic works for single precision and not for double precision where it uses the `vsqrtpd` instruction (see code [here](https://godbolt.org/z/gzXPtv)) :)